### ISSUE TITLE
How to properly handle buffers?

### ISSUE THREAD

**Issue by vfdev-5** [dmitry-vorobiev/generative-models#3]
https://github.com/dmitry-vorobiev/generative-models/blob/3d27b596be18cc38cef0ea7d20189d40b728d381/src/models/stylegan2/train.py#L18

Maybe like this:
```python

def setup_ema(ema_model, ref_model):
    # assumption: ema_model architecture == ref_model architecture
    ema_model.load_state_dict(ref_model.state_dict())
    for param in ema_model.parameters():
        param.detach_()
    # set EMA model's BN buffers as base model BN buffers:
    for m1, m2 in zip(ref_model.modules(), ema_model.modules()):
        if isinstance(m1, nn.BatchNorm2d) and isinstance(m2, nn.BatchNorm2d):
            m2.running_mean = m1.running_mean
            m2.running_var = m1.running_var
```
?

**Comment by dmitry-vorobiev** [comment]
There are no batch norms)) But G keeps moving average of disentangled latents (w_avg) to use later in inference.

From what I've seen in other works, typically these buffers are recalculated after the model has been trained by running EMA copy of G for *n* iterations in train mode with `torch.no_grad()`.

**Comment by dmitry-vorobiev** [comment]
I added a script for porting pretrained TF weights. So I can check this on converted weights later. Or maybe there is a clear answer in the official code, which I've missed out...

**Comment by vfdev-5** [comment]
> There are no batch norms)) 

True :) 
Anyway it is possible to set `w_avg` of EMA model to the same `w_avg` of source model.

> From what I've seen in other works, typically these buffers are recalculated after the model has been trained by running EMA copy of G for n iterations in train mode with torch.no_grad().

Yes, it can be an option and this is an explicit way to setup buffers but taking into account additinal execution time.

https://github.com/NVlabs/stylegan2/blob/cec605e0834de5404d5c7e5cead7053bdd0e4dde/training/networks_stylegan2.py#L194

In TF code `dlatent_avg` is non-trainable variable and probably should be cloned when ema generator is defined. Most likely it is updated as other variables...

**Comment by dmitry-vorobiev** [comment]
Looks like all non-trainable variables are copied with each EMA update:
https://github.com/NVlabs/stylegan2/blob/cec605e0834de5404d5c7e5cead7053bdd0e4dde/dnnlib/tflib/network.py#L341
They use lerp with 0.0 beta for these updates, which, if I understand correctly, should be equivalent to throwing away old values and replacing them with new ones on each step.

If this is the case, it should be easy to add :)

**Comment by dmitry-vorobiev** [comment]
Added explicit copy of *w_avg* in https://github.com/dmitry-vorobiev/generative-models/commit/65c8919207ca9ac6aca8911cc6c4c8a8e2af2554

**Comment by vfdev-5** [comment]
OK, let's close this issue as solved :)
